### PR TITLE
Add endpoint capability to lock type

### PR DESCRIPTION
**Pull request by Zitrium31**
Purpose of this change:
- add endpoint capability to lock in `fromZigbee` converter and `modernExtend` lock function
- add pincode capability for LOCK, UNLOCK, TOGGLE _state_ commands in `toZigbee lock` set converter
- add dummy garage door and lock  'GADOLOCK' device in `custom_devices_diy` definitions for test purpose. This device is made of  several endpoints using state key in `toZigbee` converter : onOff, windowCovering, lock
- fix proper toZigbee selection with device made of several endpoints using state key in `toZigbee` converter in _exposes access_ test

In short, this change supplements issue Koenkk/zigbee2mqtt#24352 for lock.

Note : A complementary PR will be submitted shortly in repo [Koenkk/zigbee2mqtt](https://github.com/Koenkk/zigbee2mqtt) to add endpoint support to lock type in lib/extension/homeassistant.ts :arrow_right: See PR https://github.com/Koenkk/zigbee2mqtt/pull/25359

:slightly_smiling_face: Thanks in advance for the review of this PR. Should you have any remark, please ask.

### CODE REVIEW

File: src/devices/custom_devices_diy.ts [hunk 1]
```diff
@@ -1193,6 +1195,18 @@ const definitions: DefinitionWithExtend[] = [
             }),
         ],
     },
+    {
+        zigbeeModel: ['GADOLOCK'],
+        model: 'GADOLOCK',
+        vendor: 'Custom devices (DiY)',
+        description: 'Garage door with lock',
```
Comment (Koenkk) [inline src/devices/custom_devices_diy.ts:1202]: Are there any instructions available on how to build such lock? This is a requirement for custom devices to be integrated.

Comment (Zitrium31) [reply src/devices/custom_devices_diy.ts:1202]: Thank you for your review.

I have actually a similar DIY device at home that controls an existing garage door motor in addition to 4 electrical latches at each corners. This project is very specific and I have my own external definition. From this point of view, I can say there is no interrest to pollute zhc repo with such device.

I found no existing device made of such lock type with endpoints. Thus I felt it was valuable (ex: to avoid regression bug like issue https://github.com/Koenkk/zigbee2mqtt/issues/24352) to add this represantative `GADOLOCK` model to re-inforce regression test on such device made of several endpoints using state key in toZigbee converter : onOff, windowCovering, lock.

This being said, I understand without any problem that this approach does not fulfill zhc policy and this device can be removed from this PR without any problem. No matter for me 🙂

File: src/converters/toZigbee.ts [hunk 2]
```diff
@@ -332,8 +332,20 @@ const converters2 = {
...
-            await entity.command('closuresDoorLock', `${value.toLowerCase()}Door`, {pincodevalue: ''}, utils.getOptions(meta.mapped, entity));
+            // If no pin code is provided, value is a only string. Ex: "UNLOCK"
+            let state = utils.isString(value) ? value.toUpperCase() : null;
+            let pincode = '';
+            // If pin code is provided, value is an object including command and code. Ex: {command: "UNLOCK", code: "1234"}
+            if (utils.isObject(value)) {
+                if (value.code) {
+                    pincode = utils.isString(value.code) ? value.code : '';
+                }
+                if (value.command) {
```
Comment (Koenkk) [inline src/converters/toZigbee.ts:343]:
```suggestion
                if (value.state) {
```

I think this is more consistent

File: src/devices/custom_devices_diy.ts (hunk 1 above)
Comment (Koenkk) [reply src/devices/custom_devices_diy.ts:1202]: Yes please remove the device from this PR

File: test/index.test.js [hunk 3]
```diff
@@ -441,7 +441,9 @@ describe('index.js', () => {
                     property = expose.property.slice(0, (expose.endpoint.length + 1) * -1);
                 }
 
-                const toZigbee = device.toZigbee.find((item) => item.key.includes(property));
+                const toZigbee = device.toZigbee.find(
```
Comment (Koenkk) [inline test/index.test.js:444]: Why was this change needed?

File: src/devices/custom_devices_diy.ts (hunk 1 above)
Comment (Zitrium31) [reply src/devices/custom_devices_diy.ts:1202]: This is understood : I gonna remove this phantom device.

File: src/converters/toZigbee.ts (hunk 2 above)
Comment (Zitrium31) [reply src/converters/toZigbee.ts:343]: Now I see your suggestion, I fully agree it is more consistent with a key named `state`. I will update that too.

File: test/index.test.js (hunk 3 above)
Comment (Zitrium31) [reply test/index.test.js:444]: This change has the same purpose as the one we did here : https://github.com/Koenkk/zigbee2mqtt/blame/bc6eccf6152c4cd91e75309f7ea0b8ad4e8835d4/lib/extension/publish.ts#L263

Let's take an example : If you have for instance a device made of a winndowClosures cluster (Ex: A garage door) and a onOff cluster (Ex: A garage light), then you have here a mix up between toZigbee converter because `find` function stops at the first key `state` found if we don't use the additional endpoint criteria. This is also explained in issue https://github.com/Koenkk/zigbee2mqtt/issues/24352.

Today, with the removal of my `GADOLOCK` device, no false positive error will be triggered (even with all others thousand devices! ).

I hope I am clear enough. Please advise if you prefer keep this change preventively or remove it because considered not relevant. At my level (a simple z2m user), I will simply follow your recommendation 🙂

File: src/devices/custom_devices_diy.ts (hunk 1 above)
Comment (Zitrium31) [reply src/devices/custom_devices_diy.ts]: Done

File: src/converters/toZigbee.ts (hunk 2 above)
Comment (Zitrium31) [reply src/converters/toZigbee.ts]: Done
